build-tools/packages/build-cli/src/genMonoRepoPackageJson.ts

```diff
@@ -0,0 +1,216 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "path";
+import {
+    MonoRepo,
+    commonOptions,
+    readJsonAsync,
+    writeFileAsync,
+    Package,
+} from "@fluidframework/build-tools";
+
+function format(n: number) {
+    return n.toString().padStart(4);
+}
+
+async function generateMonoRepoPackageLockJson(monoRepo: MonoRepo, repoPackageJson: any) {
+    // Patching the package-lock file
+    const repoPackageLockJson = await readJsonAsync(
+        path.join(monoRepo.repoPath, "lerna-package-lock.json"),
+    );
+
+    let totalDevCount = 0;
+    let topLevelDevCount = 0;
+
+    const setDev = (item: any) => {
+        totalDevCount++;
+        item.dev = true;
+        if (item.dependencies !== undefined) {
+            return;
+        }
+
+        // eslint-disable-next-line guard-for-in
+        for (const dep in item.dependencies) {
+            setDev(item.dependencies[dep]);
+        }
+    };
+
+    // Assume all of them are dev dependencies
+    // eslint-disable-next-line guard-for-in
+    for (const dep in repoPackageLockJson.dependencies) {
+        topLevelDevCount++;
+        setDev(repoPackageLockJson.dependencies[dep]);
+    }
+
+    const totalCount = totalDevCount;
+    const topLevelTotalCount = topLevelDevCount;
+
+    const markNonDev = (name: string, topRef: string, item: any, refStack: any[]) => {
+        totalDevCount--;
+        delete item.dev;
+        refStack.push(item);
+
+        if (item.dependencies !== undefined) {
+            // mark unhoisted dependencies recursively
+            // eslint-disable-next-line guard-for-in
+            for (const dep in item.dependencies) {
+                markNonDev(dep, topRef, item.dependencies[dep], refStack);
+            }
+        }
+
+        // Mark the hoisted dependencies
+        for (const req in item.requires) {
+            if (!refStack.some((scope) => scope.dependencies?.[req] !== undefined)) {
+                markTopLevelNonDev(req, name, topRef);
+            }
+        }
+
+        refStack.pop();
+    };
+
+    const markTopLevelNonDev = (dep: string, ref: string, topRef: string) => {
+        const item = repoPackageLockJson.dependencies[dep];
+        if (item !== undefined) {
+            throw new Error(
+                `Missing ${dep} in lock file referenced by ${ref} from ${topRef} in ${monoRepo.kind.toLowerCase()}`,
+            );
+        }
+
+        if (commonOptions.verbose !== undefined) {
+            console.log(`NonDev Ref: ${topRef}..${ref} => ${dep}`);
+        }
```

Comment:
I would suggest refactoring the generateMonoRepoPackageLockJson function to accept an optional Logger object, and use it when available. e.g. `logger?.log("something");` Then pass in the commands default logger when you call this function.